File: UIDevice+DisplayName.m
//  UIDevice+DisplayName.m
//
// Copyright (c) 2013 Stephan Heilner
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "UIDevice+DisplayName.h"
#include <sys/sysctl.h>

@implementation UIDevice (DisplayName)

- (NSString *)displayName {
    return [self displayNameShowSubFamily:YES];
}

- (NSString *)displayModelName {
    return [self displayNameShowSubFamily:NO];
}

- (NSString *)displayNameShowSubFamily:(BOOL)showSubFamily {
    NSString *device = [self device];

    NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
    numberFormatter.numberStyle = NSNumberFormatterDecimalStyle;
    numberFormatter.decimalSeparator = @",";

    if ([device hasPrefix:@"iPhone"]) {
        NSString *model = [device substringFromIndex:[@"iPhone" length]];
        return [self iPhoneModel:[numberFormatter numberFromString:model] showSubFamily:showSubFamily];

    } else if ([device hasPrefix:@"iPad"]) {
        NSString *model = [device substringFromIndex:[@"iPad" length]];
        return [self iPadModel:[numberFormatter numberFromString:model] showSubFamily:showSubFamily];

    } else if ([device hasPrefix:@"iPod"]) {
        NSString *model = [device substringFromIndex:[@"iPod" length]];
        return [self iPodModel:[numberFormatter numberFromString:model]];

    } else if ([device hasSuffix:@"86"] || [device isEqual:@"x86_64"]) {
        return UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone ? @"iPhone Simulator" : @"iPad Simulator";
    };

    return [self localizedModel];
}

- (NSString *)device {
    size_t size;
    sysctlbyname("hw.machine", NULL, &size, NULL, 0);
    char *machine = malloc(size);
    sysctlbyname("hw.machine", machine, &size, NULL, 0);
    NSString *device = @(machine);
    free(machine);

    return device;
}

- (NSString *)iPodModel:(NSNumber *)modelNumber {
    NSInteger intModelNumber = (NSInteger)([modelNumber floatValue] * 10);

    NSString *modelName;

    switch (intModelNumber) {
        case 11:
            modelName = @"1st Gen";
            break;
        case 21:
            modelName = @"2nd Gen";
            break;
        case 31:
            modelName = @"3rd Gen";
            break;
        case 41:
            modelName = @"4th Gen";
            break;
        case 51:
            modelName = @"5th Gen";
            break;
        default:
            modelName = @"Unknown";
            break;
    }

    return [NSString stringWithFormat:@"iPod Touch %@", modelName];
}

- (NSString *)iPadModel:(NSNumber *)modelNumber showSubFamily:(BOOL)showSubFamily {
    NSInteger intModelNumber = (NSInteger)([modelNumber floatValue] * 10);

    NSString *modelName;
    NSString *subFamily = @"";

    switch (intModelNumber) {
        case 11:
            modelName = @"1";
            break;
        case 21:
            modelName = @"2";
            subFamily = @"Wi-Fi";
            break;
        case 22:
            modelName = @"2";
            subFamily = @"GSM";
            break;
        case 23:
            modelName = @"2";
            subFamily = @"CDMA";
            break;
        case 24:
            modelName = @"2";
            subFamily = @"New";
            break;
        case 25:
            modelName = @"Mini";
            subFamily = @"Wi-Fi";
            break;
        case 26:
            modelName = @"Mini";
            subFamily = @"GSM";
            break;
        case 27:
            modelName = @"Mini";
            subFamily = @"CDMA";
            break;
        case 31:
            modelName = @"3";
            subFamily = @"Wi-Fi";
            break;
        case 32:
            modelName = @"3";
            subFamily = @"GSM";
            break;
        case 33:
            modelName = @"3";
            subFamily = @"CDMA";
            break;
        case 34:
            modelName = @"4";
            subFamily = @"Wi-Fi";
            break;
        case 35:
            modelName = @"4";
            subFamily = @"GSM";
            break;
        case 36:
            modelName = @"4";
            subFamily = @"GSM+CDMA";
            break;
        case 41:
            modelName = @"Air";
            subFamily = @"Wi-Fi";
            break;
        case 42:
            modelName = @"Air";
            subFamily = @"Cellular";
            break;
        case 43:
            modelName = @"Air";
            subFamily = @"China";
            break;
        case 44:
            modelName = @"Mini 2";
            subFamily = @"Wi-Fi";
            break;
        case 45:
            modelName = @"Mini 2";
            subFamily = @"Cellular";
            break;
        case 46:
            modelName = @"Mini 2";
            subFamily = @"China";
            break;
        case 47:
            modelName = @"Mini 3";
            subFamily = @"Wi-Fi";
            break;
        case 48:
            modelName = @"Mini 3";
            subFamily = @"Cellular";
            break;
        case 53:
            modelName = @"Air 2";
            subFamily = @"Wi-Fi";
            break;
        case 54:
            modelName = @"Air 2";
            subFamily = @"Cellular";
            break;
        default:
            modelName = @"Unknown";
            break;
    }

    if (showSubFamily && [subFamily length] > 0) {
        return [NSString stringWithFormat:@"iPad %@ %@", modelName, subFamily];
    }
    
    return [NSString stringWithFormat:@"iPad %@", modelName];
}

- (NSString *)iPhoneModel:(NSNumber *)modelNumber showSubFamily:(BOOL)showSubFamily {
    NSInteger intModelNumber = (NSInteger)([modelNumber floatValue] * 10);

    NSString *modelName;
    NSString *subFamily = @"";

    switch (intModelNumber) {
        case 11:
            modelName = @"2G";
            break;
        case 12:
            modelName = @"3G";
            break;
        case 21:
            modelName = @"3GS";
            break;
        case 31:
            modelName = @"4";
            subFamily = @"GSM";
            break;
        case 32:
            modelName = @"4";
            subFamily = @"8GB";
            break;
        case 33:
            modelName = @"4";
            subFamily = @"CDMA";
            break;
        case 41:
            modelName = @"4S";
            break;
        case 51:
            modelName = @"5";
            subFamily = @"GSM";
            break;
        case 52:
            modelName = @"5";
            subFamily = @"GSM+CDMA";
            break;
        case 53:
            modelName = @"5c";
            subFamily = @"GSM";
            break;
        case 54:
            modelName = @"5c";
            subFamily = @"GSM+CDMA";
            break;
        case 61:
            modelName = @"5s";
            subFamily = @"GSM";
            break;
        case 62:
            modelName = @"5s";
            subFamily = @"GSM+CDMA";
            break;
        case 71:
            modelName = @"6 Plus";
            break;
        case 72:
            modelName = @"6";
            break;
        default:
            modelName = @"Unknown";
            break;
    }

    if (showSubFamily && [subFamily length] > 0) {
        return [NSString stringWithFormat:@"iPhone %@ %@", modelName, subFamily];
    }
    
    return [NSString stringWithFormat:@"iPhone %@", modelName];
}

@end

File: UIDevice+DisplayName.swift
// Copyright (c) 2018 Stephan Heilner
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import Foundation

private class DummyClassToGetBundle {}

public extension UIDevice {
    
    private var lastUpdated: Date? {
        get {
            let timeInterval = UserDefaults.standard.double(forKey: "UIDevice+DisplayName-LastUpdated")
            return timeInterval > 0 ? Date(timeIntervalSince1970: timeInterval) : nil
        }
        set {
            if let timeInterval = newValue?.timeIntervalSince1970 {
                UserDefaults.standard.set(timeInterval, forKey: "UIDevice+DisplayName-LastUpdated")
            } else {
                UserDefaults.standard.removeObject(forKey: "UIDevice+DisplayName-LastUpdated")
            }
        }
    }
    
    private static var isUpdating = false
    
    private static let devicesFileURL: URL? = {
        guard let devicesFileURL = FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).last?.appendingPathComponent("UIDevice").appendingPathComponent("devices.json")
        else { return nil }
        
        if !FileManager.default.fileExists(atPath: devicesFileURL.path) {
            let localFileURL = Bundle(for: DummyClassToGetBundle.self).url(forResource: "devices", withExtension: "json")
            do {
                try FileManager.default.createDirectory(at: devicesFileURL.deletingLastPathComponent(), withIntermediateDirectories: true, attributes: nil)
                if let localFileURL = localFileURL {
                    try FileManager.default.copyItem(at: localFileURL, to: devicesFileURL)
                    if let creationDate = DateFormatter.rfc.date(from: "Sun, 01 Nov 2020 00:00:00 GMT") {
                        try FileManager.default.setAttributes([.creationDate: creationDate], ofItemAtPath: devicesFileURL.path)
                    }
                }
            } catch {
                print("Error copying devices.json file", error)
                return localFileURL
            }
        }
        
        return devicesFileURL
    }()
    
    private static var deviceTypes: [String: String] = {
        guard let devicesFileURL = devicesFileURL
        else { return [:] }
        
        do {
            let jsonData = try Data(contentsOf: devicesFileURL)
            return try (JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any])?["deviceTypes"] as? [String: String] ?? [:]
        } catch {
            print("Unable to load devices", error)
            return [:]
        }
    }()
    
    private static var devices: [String: Any] = {
        guard let devicesFileURL = devicesFileURL else { return [:] }
        
        do {
            let jsonData = try Data(contentsOf: devicesFileURL)
            return (try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any])?["devices"] as? [String: Any] ?? [:]
        } catch {
            print("Unable to load devices", error)
            return [:]
        }
    }()
    
    func displayName(includeType: Bool = true, deviceIdentifier: String? = nil) -> String {
        checkForUpdates()
        
        switch deviceIdentifier ?? self.deviceIdentifier() {
        case let simulator where simulator.hasPrefix("x86"):
            switch userInterfaceIdiom {
            case .unspecified:
                return "Unspecified Simulator"
            case .phone:
                return "iPhone Simulator"
            case .pad:
                return "iPad Simulator"
            case .tv:
                return "Apple TV Simulator"
            case .carPlay:
                return "CarPlay Simulator"
            case .mac:
                return "Mac Simulator"
            @unknown default:
                return "Unspecified Simulator"
            }
        case let device:
            for (prefix, name) in UIDevice.deviceTypes {
                guard let range = device.range(of: prefix) else { continue }
                
                return displayName(for: name, prefix: prefix, model: String(device[range.upperBound...]), includeType: includeType)
            }
            return device
        }
    }
    
    private func displayName(for deviceName: String, prefix: String, model: String, includeType: Bool) -> String {
        let unknownDevice = "Unknown \(deviceName)"
        
        guard let devices = UIDevice.devices[deviceName] as? [[String: Any]] else { return unknownDevice }
        
        var name: String?
        var type: String?
        
        for device in devices {
            guard let deviceIdentifier = device.keys.first, deviceIdentifier == "\(prefix)\(model)" else { continue }
            
            for (key, value) in device.values.first as? [String: String] ?? [:] {
                switch key {
                case "name":
                    name = value
                case "type":
                    type = value
                default: ()
                }
            }
        }
        
        if let name = name {
            if includeType, let type = type {
                return "\(deviceName) \(name) (\(type))"
            }
            return "\(deviceName) \(name)"
        }
        return unknownDevice
    }
    
    private func deviceIdentifier() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        return machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0
            else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
    }
    
    private func checkForUpdates() {
        guard !UIDevice.isUpdating,
              let jsonURL = URL(string: "https://cdn.churchofjesuschrist.org/mobile/devices.json")
        else { return }
        
        if let lastUpdated = lastUpdated, Date().timeIntervalSince(lastUpdated) < 604800 {
            // Only check once every week (604800 seconds) at most
            return
        }
        
        var request = URLRequest(url: jsonURL)
        if let fileURL = UIDevice.devicesFileURL, let date = try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.creationDate] as? Date {
            request.setValue(DateFormatter.rfc.string(from: date), forHTTPHeaderField: "If-Modified-Since")
        }
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            let response = response as? HTTPURLResponse
            do {
                if let response = response, response.statusCode == 304 {
                    // Do nothing, file hasn't changed
                    self?.lastUpdated = Date()
                } else if let error = error {
                    print("Unable to update devices.json", error)
                } else if let data = data, let devicesFileURL = UIDevice.devicesFileURL {
                    let json = try JSONSerialization.jsonObject(with: data, options: [])
                    try data.write(to: devicesFileURL, options: .atomic)
                    
                    if let lastModifiedString = response?.allHeaderFields["Last-Modified"] as? String, let modifiedDate = DateFormatter.rfc.date(from: lastModifiedString) {
                        try? FileManager.default.setAttributes([.creationDate: modifiedDate], ofItemAtPath: devicesFileURL.path)
                    }
                    
                    if let deviceTypes = (json as? [String: Any])?["deviceTypes"] as? [String: String] {
                        UIDevice.deviceTypes = deviceTypes
                    }
                    if let devices = (json as? [String: Any])?["devices"] as? [String: Any] {
                        UIDevice.devices = devices
                    }
                    
                    self?.lastUpdated = Date()
                }
            } catch {
                print("Unable to parse/save devices.json", error)
            }
            UIDevice.isUpdating = false
        }.resume()
    }
    
}

private extension DateFormatter {
    static let rfc: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter
    }()
}
